common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java

```diff
@@ -49,17 +71,85 @@ public abstract class ClientPacketListenerMixin {
     @Shadow
     protected abstract void send(ServerboundResourcePackPacket.Action action);
 
+    @Shadow
+    public abstract void send(Packet<?> packet);
+
+    @Shadow
+    private CommandDispatcher<SharedSuggestionProvider> commands;
+
+    @Shadow
+    @Final
+    private Minecraft minecraft;
+
+    @Shadow
+    public abstract FeatureFlagSet enabledFeatures();
+
+    @Shadow
+    private LayeredRegistryAccess<ClientRegistryLayer> registryAccess;
+
+    @Shadow
+    private MessageSignatureCache messageSignatureCache;
+
+    @Shadow
+    private SignedMessageChain.Encoder signedMessageEncoder;
+
+    @Shadow
+    private LastSeenMessagesTracker lastSeenMessages;
+
     private static boolean isRenderThread() {
-        return (McUtils.mc().isSameThread());
+        return McUtils.mc().isSameThread();
+    }
+
+    @Inject(method = "sendChat", at = @At("HEAD"), cancellable = true)
+    private void onChatPre(String string, CallbackInfo ci) {
+        ChatSentEvent result = EventFactory.onChatSent(string);
+        if (result.isCanceled()) {
+            ci.cancel();
+        }
+
+        if (!Objects.equals(string, result.getMessage())) {
+            // Note: This code is taken from the MC source code. I don't think there is a sane way of not doing this,
+            // unfortunately (in 1.19.2, there was, not in 1.19.3).
+            //       For anyone concerned, this does not mess with NoChatReport at the time of writing.
+
+            Instant instant = Instant.now();
```

Comment:
This sounds a bit scary, too. Can we end up signing messages, claiming the user sent them, when they did in fact not, since we modified the message? I can't think of really how this can be mis-used, but the entire chat report thing is creepy, and I'd hate for us to make anything stupid here.

Reply:
I checked where and why we need to change the sent message (cancelling is another story, that does not play bad with chat reporting). It is done in two places: in CommandAliasesFeature and TradeMarketPriceConversionFeature. As I said in discord, I wonder if the first one is broken. For the second one, I've been thinking for a long time to write a GUI to the trade market so you can type the search terms in the GUI, not the chat. If we do that, we can do price substitution in the GUI string, even before it is sent in chat.

If it ends up that we don't need to support changing the message text, I think we should remove this code.

Reply:
My latest commit to this branch (and the orignal port) should resolve your concerns.